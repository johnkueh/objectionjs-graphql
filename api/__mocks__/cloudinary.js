const cloudinary = jest.genMockFromModule('cloudinary');
export default cloudinary;
