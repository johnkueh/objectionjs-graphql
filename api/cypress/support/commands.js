Cypress.Commands.add('deleteUser', email => {
  return cy.task('deleteUser', email);
});

Cypress.Commands.add('factory', args => {
  return cy.task('factory', args);
});

Cypress.Commands.add('createUserAndLogin', args => {
  cy.task('factory', {
    method: 'create',
    type: 'user',
    args,
    include: ['jwt']
  }).then(user => {
    cy.setCookie('jwt', user.jwt);
    cy.wrap(user);
  });
});

Cypress.Commands.add('createUserWithWorkspacesAndLogin', args => {
  cy.task('factory', {
    method: 'create',
    type: 'userWithWorkspace',
    args,
    include: ['jwt', 'workspaces']
  }).then(user => {
    cy.setCookie('jwt', user.jwt);
    cy.wrap(user);
  });
});
