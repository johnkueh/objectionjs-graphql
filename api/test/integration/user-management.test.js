import '../support/transactional-tests';
import '../factories';
import cloudinary from 'cloudinary';
import factory from 'factory-girl';
import { query } from '../support/apollo-test-helper';
import User from '../../models/user';

describe('Fetching user profile', () => {
  let user;
  const ME = `
    query {
      me {
        id
        name
        email
        logo {
          id
          publicId
        }
      }
    }
  `;

  beforeEach(async () => {
    user = await factory.create('user', {
      email: 'john@doe.com',
      password: 'password'
    });
  });

  it('is not able to fetch user profile without credentials', async () => {
    const res = await query({
      query: ME
    });

    expect(res.errors[0].extensions).toMatchSnapshot();
  });

  it('is not able to fetch user profile with wrong credentials', async () => {
    const res = await query({
      query: ME
    });

    expect(res.errors[0].extensions).toMatchSnapshot();
  });

  it('is able to fetch user profile with credentials', async () => {
    const res = await query({
      context: { user },
      query: ME
    });

    expect(res).toMatchSnapshot({
      data: {
        me: {
          id: expect.any(String),
          name: user.name,
          email: user.email,
          logo: null
        }
      }
    });
  });

  it('is able to fetch user profile with logo', async () => {
    await factory.create('image', {
      publicId: 'user-logo-id',
      imageableType: 'UserLogo',
      imageableId: user.id
    });

    const res = await query({
      context: { user },
      query: ME
    });

    expect(res).toMatchSnapshot({
      data: {
        me: {
          id: expect.any(String),
          name: user.name,
          email: user.email,
          logo: {
            id: expect.any(String),
            publicId: 'user-logo-id'
          }
        }
      }
    });
  });
});

describe('Updating user profile', () => {
  let user;
  const UPDATE_USER = `
    mutation($input: UpdateUserInput!) {
      updateUser(input: $input) {
        id
        name
        email
      }
    }
  `;

  beforeEach(async () => {
    user = await factory.create('user', {
      email: 'john@doe.com',
      password: 'password'
    });
  });

  it('can update user profile with valid fields', async () => {
    const res = await query({
      context: { user },
      query: UPDATE_USER,
      variables: {
        input: {
          email: 'darth@vader.com',
          name: 'Darth Vader'
        }
      }
    });

    expect(res).toMatchSnapshot({
      data: {
        updateUser: {
          id: expect.any(String),
          name: 'Darth Vader',
          email: 'darth@vader.com'
        }
      }
    });
  });

  it('can update user password', async () => {
    await query({
      context: { user },
      query: UPDATE_USER,
      variables: {
        input: {
          password: 'newpassword'
        }
      }
    });

    const updatedUser = await User.query().findById(user.id);
    expect(updatedUser.validPassword('newpassword')).toBe(true);
  });

  it('fails to update with invalid fields', async () => {
    const res = await query({
      context: { user },
      query: UPDATE_USER,
      variables: {
        input: {
          name: '',
          email: 'hel@per',
          password: 'abc'
        }
      }
    });

    expect(res.errors[0].extensions).toMatchSnapshot();
  });

  describe('Managing user logo', () => {
    const DELETE_IMAGE = `
      mutation($input: DeleteImageInput!) {
        deleteImage(input: $input) {
          count
        }
      }
    `;
    it('is not able to delete logo owned by other user', async () => {
      const image = await factory.create('image', {
        publicId: 'existing-public-id',
        imageableType: 'UserLogo',
        imageableId: 'other-user-id'
      });

      const res = await query({
        context: { user },
        query: DELETE_IMAGE,
        variables: {
          input: {
            id: image.id
          }
        }
      });

      expect(res.errors[0].extensions).toMatchSnapshot();
      expect(cloudinary.uploader.destroy).not.toHaveBeenCalled();
    });

    it('is able to delete logo owned by user', async () => {
      const image = await factory.create('image', {
        publicId: 'existing-public-id',
        imageableType: 'UserLogo',
        imageableId: user.id
      });

      const res = await query({
        context: { user },
        query: DELETE_IMAGE,
        variables: {
          input: {
            id: image.id
          }
        }
      });

      expect(res).toMatchSnapshot({
        data: {
          deleteImage: {
            count: 1
          }
        }
      });

      const userLogo = await user.$relatedQuery('logo');
      expect(userLogo).toBeUndefined();

      expect(cloudinary.uploader.destroy).toHaveBeenCalledWith(image.publicId);
    });
  });
});
