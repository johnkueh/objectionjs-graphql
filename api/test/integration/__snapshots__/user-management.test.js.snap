// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fetching user profile is able to fetch user profile with credentials 1`] = `
Object {
  "data": Object {
    "me": Object {
      "email": "john@doe.com",
      "id": Any<String>,
      "logo": null,
      "name": "John Doe",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Fetching user profile is able to fetch user profile with logo 1`] = `
Object {
  "data": Object {
    "me": Object {
      "email": "john@doe.com",
      "id": Any<String>,
      "logo": Object {
        "id": Any<String>,
        "publicId": "user-logo-id",
      },
      "name": "John Doe",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Fetching user profile is not able to fetch user profile with wrong credentials 1`] = `
Object {
  "code": "BAD_USER_INPUT",
  "exception": Object {
    "errors": Object {
      "auth": "You are not authorized to perform this action",
    },
  },
}
`;

exports[`Fetching user profile is not able to fetch user profile without credentials 1`] = `
Object {
  "code": "BAD_USER_INPUT",
  "exception": Object {
    "errors": Object {
      "auth": "You are not authorized to perform this action",
    },
  },
}
`;

exports[`Updating user profile Managing user logo is able to delete logo owned by user 1`] = `
Object {
  "data": Object {
    "deleteImage": Object {
      "count": 1,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Updating user profile Managing user logo is not able to delete logo owned by other user 1`] = `
Object {
  "code": "BAD_USER_INPUT",
  "exception": Object {
    "errors": Object {
      "auth": "You are not authorized to perform this action",
    },
  },
}
`;

exports[`Updating user profile can update user profile with valid fields 1`] = `
Object {
  "data": Object {
    "updateUser": Object {
      "email": "darth@vader.com",
      "id": Any<String>,
      "name": "Darth Vader",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Updating user profile fails to update with invalid fields 1`] = `
Object {
  "code": "BAD_USER_INPUT",
  "exception": Object {
    "errors": Object {
      "email": "Email must be a valid email",
      "name": "Name must be at least 1 characters",
      "password": "Password must be at least 6 characters",
    },
  },
}
`;
